proxy/stats_handler.go

```diff
@@ -0,0 +1,215 @@
+package proxy
+
+import (
+	"fmt"
+	"net/http"
+	"strings"
+	"sync"
+	"time"
+
+	log "github.com/Sirupsen/logrus"
+	jwt "github.com/dgrijalva/jwt-go"
+	"github.com/gorilla/websocket"
+
+	"github.com/rancherio/websocket-proxy/common"
+)
+
+type StatsHandler struct {
+	backend         backendProxy
+	parsedPublicKey interface{}
+}
+
+func (h *StatsHandler) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
+	multiHost := false
+
+	if strings.HasSuffix(req.URL.Path, "project/") || strings.HasSuffix(req.URL.Path, "service/") {
+		multiHost = true
+	}
+
+	hostKeys, urls, authed := h.auth(req, multiHost)
+	if !authed {
+		return
+	}
+
+	upgrader := websocket.Upgrader{
+		CheckOrigin: func(r *http.Request) bool { return true },
+	}
+	ws, err := upgrader.Upgrade(rw, req, nil)
+	if err != nil {
+		http.Error(rw, "Failed to upgrade connection.", 500)
+		return
+	}
+
+	msgKeys := make([]string, len(hostKeys))
+	respChannels := make([]<-chan common.Message, len(hostKeys))
+
+	var mutex sync.Mutex
+	var countMutex sync.Mutex
+
+	doneCounter := len(hostKeys)
+
+	for i, hostKey := range hostKeys {
+		var err error
+		msgKeys[i], respChannels[i], err = h.backend.initializeClient(hostKey)
+		if err != nil {
+			h.closeConnection(ws)
+			return
+		}
+
+		// Send response messages to client
+		go func(pos int) {
+			for {
+				message, ok := <-respChannels[pos]
+				if !ok {
+					return
+				}
+				switch message.Type {
+				case common.Body:
+					ws.SetWriteDeadline(time.Now().Add(10 * time.Second))
+					mutex.Lock()
+					if err := ws.WriteMessage(1, []byte(message.Body)); err != nil {
+						for j, hostKey := range hostKeys {
+							h.backend.closeConnection(hostKey, msgKeys[j])
+						}
+						h.closeConnection(ws)
+					}
+					mutex.Unlock()
+				case common.Close:
+					countMutex.Lock()
+					doneCounter--
+					if doneCounter == 0 {
+						h.closeConnection(ws)
+					}
+					h.backend.closeConnection(hostKeys[pos], msgKeys[pos])
+					countMutex.Unlock()
+				}
+			}
+		}(i)
+
+		if err = h.backend.connect(hostKey, msgKeys[i], urls[i]); err != nil {
+			fmt.Println("Error connecting to backend, err = ", err)
+			return
+		}
+	}
+
+}
```

Comment:
@wlan0 The backend connections are staying open too long after the client closes the ws connection. I believe this is because you do not handle close messages from the client and instead rely on the underlying socket connection to be closed.

We need to read all messages coming from the client and if we receive a close message, we need to close the websocket connection on our end as well. If you add something like this after the for loop that spins up the goroutine, it should fix the problem:
```
	for {
		_, _, err := ws.ReadMessage()
		if err != nil {
			h.closeConnection(ws)
			return
		}
	}
```
FYI, in the library we're using, an error is returned from ReadMessage if the client has send a close message.